// @astrojs/vercel/dist/edge/adapter.js
import { relative as relativePath } from "node:path";
import { fileURLToPath } from "node:url";
import {
  copyFilesToFunction,
  getFilesFromFolder,
  getVercelOutput,
  removeDir,
  writeJson
} from "../lib/fs.js";
import { getRedirects } from "../lib/redirects.js";
const PACKAGE_NAME = "@astrojs/vercel/edge";
function getAdapter() {
  return {
    name: PACKAGE_NAME,
    serverEntrypoint: `${PACKAGE_NAME}/entrypoint`,
    exports: ["default"]
  };
}
function vercelEdge({ includeFiles = [] } = {}) {
  let _config;
  let buildTempFolder;
  let functionFolder;
  let serverEntry;
  let needsBuildConfig = false;
  return {
    name: PACKAGE_NAME,
    hooks: {
      "astro:config:setup": ({ config, updateConfig }) => {
        needsBuildConfig = !config.build.client;
        const outDir = getVercelOutput(config.root);
        updateConfig({
          outDir,
          build: {
            serverEntry: "entry.mjs",
            client: new URL("./static/", outDir),
            server: new URL("./dist/", config.root)
          }
        });
      },
      "astro:config:done": ({ setAdapter, config }) => {
        setAdapter(getAdapter());
        _config = config;
        buildTempFolder = config.build.server;
        functionFolder = new URL("./functions/render.func/", config.outDir);
        serverEntry = config.build.serverEntry;
        if (config.output === "static") {
          throw new Error(`
		[@astrojs/vercel] \`output: "server"\` is required to use the edge adapter.
	
	`);
        }
      },
      "astro:build:start": ({ buildConfig }) => {
        if (needsBuildConfig) {
          buildConfig.client = new URL("./static/", _config.outDir);
          buildTempFolder = buildConfig.server = new URL("./dist/", _config.root);
          serverEntry = buildConfig.serverEntry = "entry.mjs";
        }
      },
      "astro:build:setup": ({ vite, target }) => {
        var _a;
        if (target === "server") {
          vite.resolve || (vite.resolve = {});
          (_a = vite.resolve).alias || (_a.alias = {});
          const aliases = [{ find: "react-dom/server", replacement: "react-dom/server.browser" }];
          if (Array.isArray(vite.resolve.alias)) {
            vite.resolve.alias = [...vite.resolve.alias, ...aliases];
          } else {
            for (const alias of aliases) {
              vite.resolve.alias[alias.find] = alias.replacement;
            }
          }
          vite.ssr = {
            target: "webworker",
            noExternal: true
          };
          vite.build || (vite.build = {});
          vite.build.minify = true;
        }
      },
      "astro:build:done": async ({ routes }) => {
        const entry = new URL(serverEntry, buildTempFolder);
        const generatedFiles = await getFilesFromFolder(buildTempFolder);
        const commonAncestor = await copyFilesToFunction(
          [...generatedFiles, ...includeFiles.map((file) => new URL(file, _config.root))],
          functionFolder
        );
        await removeDir(buildTempFolder);
        await writeJson(new URL(`./.vc-config.json`, functionFolder), {
          runtime: "edge",
          entrypoint: relativePath(commonAncestor, fileURLToPath(entry))
        });
        await writeJson(new URL(`./config.json`, _config.outDir), {
          version: 3,
          routes: [
            ...getRedirects(routes, _config),
            { handle: "filesystem" },
            { src: "/.*", dest: "render" }
          ]
        });
      }
    }
  };
}
export {
  vercelEdge as default
};
